crates/rpc/rpc/src/eth/api/server.rs

```diff
@@ -180,11 +180,44 @@ where
 
     async fn fee_history(
         &self,
-        _block_count: U256,
-        _newest_block: BlockNumber,
+        block_count: u64,
+        newest_block: BlockNumber,
         _reward_percentiles: Option<Vec<f64>>,
     ) -> Result<FeeHistory> {
-        Err(internal_rpc_err("unimplemented"))
+        if block_count == 0 {
+            return Ok(FeeHistory::default())
+        }
+
+        let headers = self
+            .inner
+            .client
+            .headers_range(
+                BlockHashOrNumber::Number(newest_block - block_count)..(newest_block + 1).into(),
+            )
+            .to_rpc_result()?;
```

Comment:
this works, but in order to make this really performant we should cache all of that because this gets called a lot and all of that data is available and relatively cheap to keep in memory (for a reasonable range of blocks).

ref https://github.com/foundry-rs/foundry/blob/master/anvil/src/eth/fees.rs#L161

Reply:
makes sense. You mean caching in some in-memory data structure and not persisting in a separate db table, right?

Reply:
yeah storing the rewards,basefee per block based on a configured percentile and then calculating the `FeeHistory` based on the requested block range
tbh I can't really remember how the math works, but the cache gets populated when a new block is executed and then the rpc handler loops over the requested block range, reads the cache items and calculates the response.

https://github.com/foundry-rs/foundry/blob/550c548a7aa94a45c1f448071307108867848429/anvil/src/eth/fees.rs#L319-L327

Reply:
oh interesting, i thought of populating the cache on `feeHistory` request, so if the node user doesn't use fee-related RPC methods, reth doesn't store it in memory and keeps the footprint lower.

Reply:
but maybe you're right, since `eth_feeHistory` supports only up to 1024 blocks, we can keep the cache for these `latest-1024..latest` blocks and alter it on every block execution

Reply:
> I believe geth calls the type that provides this info an oracle in the codebase

that's right, and as I see they're querying the database on demand and not populating the cache with every block execution: https://github.com/ethereum/go-ethereum/blob/77380b9629e4f679778ce9de037ffdaf9c925238/eth/gasprice/feehistory.go#L223.

Also, they have an LRU cache for 2048 entries: https://github.com/ethereum/go-ethereum/blob/77380b9629e4f679778ce9de037ffdaf9c925238/eth/gasprice/gasprice.go#L118

Reply:
for the `reward`, I see two potential options for caching:
1. As in Anvil, store pre-calculated percentiles with a 0.5 resolution
2. Store raw transaction rewards and calculate percentiles on-demand during RPC request